### PR TITLE
split api into routes.go issue #2391

### PR DESCRIPTION
**Pull request by petabytestorage**
https://github.com/NebulousLabs/Sia/issues/2391

### CODE REVIEW

File: api/routes.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+	"time"
+
+	"github.com/NebulousLabs/Sia/build"
+	"github.com/julienschmidt/httprouter"
+)
+
+// buildHttpRoutes sets up and returns an * httprouter.Router.
+// it connected the Router to the given api using the required
+// parameters: requiredUserAgent and requiredPassword
+func buildHttpRoutes(api *API, requiredUserAgent string, requiredPassword string) *httprouter.Router {
```
Comment (DavidVorick) [inline api/routes.go]: make this a function of the API instead of passing in the api as an argument, otherwise it's not clear that the state of the api is being modified

File: api/routes.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+			return
+		}
+		h(w, req, ps)
+	}
+}
+
+// cleanCloseHandler wraps the entire API, ensuring that underlying conns are
+// not leaked if the rmeote end closes the connection before the underlying
+// handler finishes.
+func cleanCloseHandler(next http.Handler) http.Handler {
```
Comment (DavidVorick) [inline api/routes.go]: per convention, all lowercase functions (non-exported) are put in the file before all upper case functions

Comment (avahowell) [inline api/routes.go]: Should this be defined in `routes.go`? Seem orthogonal to route definition, as do some of the other funcs (`RequireUserAgent` etc)

File: api/routes.go (hunk 1 above)
Comment (avahowell) [reply api/routes.go]: Alternatively, we could change this func to be stateless and return the router, and assign the API's `router` to the value returned by the func where the API is initialized.

File: api/routes.go (hunk 2 above)
Comment (petabytestorage) [reply api/routes.go]: done @DavidVorick

File: api/routes.go (hunk 1 above)
Comment (petabytestorage) [reply api/routes.go]: all good points and either one is a step up in terms of clarity but i would not know which is better.

File: api/routes.go (hunk 2 above)
Comment (petabytestorage) [reply api/routes.go]: i don't mind wherever it goes and it seems like it should stick with the `api.router = ` piece @avahowell

File: api/routes_test.go [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package api
+
+import (
+	"testing"
+)
+
+func TestBuildHttpRoutes(t *testing.T) {
```
Comment (DavidVorick) [inline api/routes_test.go]: All functions, including test functions, need a docstring. This one seems a bit odd to me, because I could imagine adjusting the api object and buildRoutes method such that just calling buildHttpRoutes would cause a panic. The full set of routes get built every time you call `newServerTester`, so this might be a situation where we don't need this test.

Comment (petabytestorage) [reply api/routes_test.go]: doc strings and tests should come after design decisions. i don't know where `api.router =` and cleanCloseHandler etc should live so I don't know how to document / improve this test. i agree it may likely be removed. so my suggestion is the lead core devs take this over and arbitrate.

Comment (avahowell) [reply api/routes_test.go]: I think `api.router =` and `cleanCloseHandler` etc should live in `api.go`, because it's where the API is initialized. `routes.go` constructs the router, `api.go` constructs the API object.

File: api/routes.go (hunk 1 above)
Comment (DavidVorick) [reply api/routes.go]: Needs to be a method of the api I believe because when you build the routes you need to understand which modules have been initialized.

**Comment by DavidVorick** [comment]
I am going to disagree with @avahowell about where to put the functions, I think they make sense where they have been placed in this PR because they have been put in the only files where they are used. They are middlewares that wrap the router, therefore that's a good place to have them.

The create router function should be a function of the api, since it is adding a router to the api. It probably should not return the router, since the api router gets set inside of the function, if you need it you can reference the api's router. So we can strip the return value I think.

I don't think we need the test. I'm not opposed to keeping it, but pretty much every other test we have also initializes the api router in some form, if there's a problem with initialization a test will definitely catch it.

This PR is almost ready to go.

**Comment by fnvhash** [comment]
thanks @DavidVorick  all done i believe now
